### PR TITLE
Remove TKey : notnull constraint from `IDictionary<TKey, TValue>`

### PR DESCRIPTION
**Pull request by stephentoub**
The interface explicitly documents that some implementations may choose to support null keys.

Fixes https://github.com/dotnet/corefx/issues/42386
cc: @buyaa-n, @safern, @sharwell

### CODE REVIEW

**Review by safern — COMMENTED** [review]
What about IReadOnlyDictionary and others?

https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/IReadOnlyDictionary.cs#L10

**Comment by stephentoub** [comment]
> What about IReadOnlyDictionary and others?

Yes, we should make the same change for IReadOnlyDictionary.  Good catch.

What "others"?

**Comment by safern** [comment]
> What "others"?

I don’t think there are others.

**Comment by safern** [comment]
Actually I found some extension methos on `IDictionary` and `IReadOnlyDictionary` that have that constraint, should we remove it for those? Sorry I was on the phone earlier and couldn't do a good search on the repo.

https://github.com/dotnet/runtime/blob/master/src/libraries/System.Collections/ref/System.Collections.cs#L48
https://github.com/dotnet/runtime/blob/master/src/libraries/System.Collections/ref/System.Collections.cs#L51
https://github.com/dotnet/runtime/blob/master/src/libraries/System.Collections/ref/System.Collections.cs#L52

Those are all I could find.

**Comment by stephentoub** [comment]
Thanks.  I searched as well and pushed an update.

File: src/libraries/System.Collections/src/System/Collections/Generic/CollectionExtensions.cs [hunk 1]
```diff
@@ -9,13 +9,13 @@ namespace System.Collections.Generic
     public static class CollectionExtensions
     {
         [return: MaybeNull]
-        public static TValue GetValueOrDefault<TKey, TValue>(this IReadOnlyDictionary<TKey, TValue> dictionary, TKey key) where TKey : notnull
+        public static TValue GetValueOrDefault<TKey, TValue>(this IReadOnlyDictionary<TKey, TValue> dictionary, TKey key)
         {
             return dictionary.GetValueOrDefault(key, default);
         }
 
         [return: MaybeNull]
```
Comment (sharwell) [inline src/libraries/System.Collections/src/System/Collections/Generic/CollectionExtensions.cs]: Nice. This needs to be `MaybeNullIfNull("defaultValue")`, but that doesn't exist.

**Review by sharwell — APPROVED** [review]

**Review by safern — APPROVED** [review]

**Pull request merged**